support prefill cuda graph

ZhouMengLei1999 opened the pull request:
## Motivation
Added support for prefill CUDA graphs
## Modifications
1. Specified through the parameter --enable-prefill-cuda-graph
2. while using --cuda-graph-prefill-max-seqlen to limit the maximum seqlen that the prefill CUDA graph can function on, length below the maximum value will be padded to a multiple of 32. 
3. Currently, only flashinfer attention backend is supported

ZhouMengLei1999 committed:
support prefill cuda graph

fzyzcjy commented:
Hi, is this ready for review, or is it still WIP?

ZhouMengLei1999 commented:
> Hi, I would appreciate it if I could know is this ready for review, or is it still WIP?

Yeah, I just performed rebase and fix some bugs, but why I didn't find CI/CD for my pr?

fzyzcjy commented:
b/c after a change you need some maintainers to trigger ci...

ZhouMengLei1999 commented:
> b/c after a change you need some maintainers to trigger ci...
Hi, I just fix some bugs of this pr, can you help to tigger ci?

Qiaolin-Yu commented:
> > b/c after a change you need some maintainers to trigger ci...
> 
> Does everytime when I change this pr, need some maintainers to trigger ci?

Yes.

fzyzcjy commented on python/sglang/srt/server_args.py:
```diff
@@ -501,6 +503,9 @@ def print_deprecated_warning(message: str):
             ):
                 # use bf16 for mxfp4 triton kernels
                 self.dtype = "bfloat16"
+        if self.enable_prefill_cuda_graph == True:
```
oh yes this is what I want to say as well (change all `a == True` to `a`)

fzyzcjy commented on python/sglang/srt/server_args.py:
```diff
@@ -508,6 +510,10 @@ def print_deprecated_warning(message: str):
 
+        if self.enable_prefill_cuda_graph == True:
+            self.disable_radix_cache = True
+            self.attention_backend = "flashinfer"
```
nit: maybe add a logger message to tell users we auto do these several configurations

fzyzcjy commented on python/sglang/srt/server_args.py:
```diff
@@ -1684,6 +1690,16 @@ def add_cli_args(parser: argparse.ArgumentParser):
+        parser.add_argument(
+            "--cuda-graph-prefill-max-seqlen",
+            type=int,
+            help="Set the prefill seqlen eql range(32, cuda_graph_prefill_max_seqlen, 32) for cuda graph.",
```
tiny nit

```suggestion
            help="Set the prefill seqlen range(32, cuda_graph_prefill_max_seqlen, 32) for cuda graph.",
```

fzyzcjy commented on python/sglang/srt/model_executor/forward_batch_info.py:
```diff
@@ -127,7 +127,8 @@ def is_extend_or_draft_extend_or_mixed(self):
     def is_cuda_graph(self):
         return (
-            self == ForwardMode.DECODE
+            self == ForwardMode.EXTEND
```
qq: will this cause trouble when the prefill cuda graph is not enabled

fzyzcjy commented on python/sglang/srt/model_executor/cuda_graph_runner.py:
```diff
@@ -383,6 +417,150 @@ def __init__(self, model_runner: ModelRunner):
+            else max_decode_num_token
+        )
+
+        self.model_runner.attn_backend.init_cuda_graph_state(
```
hmm it seems we init_cuda_graph_state twice: once at GraphRunner.init, another at capture_prefill. Is it possible we only init it once at the GraphRunner.init function, since the new code looks asymmetric and the "init things twice" may cause confusion.

fzyzcjy commented on python/sglang/srt/layers/attention/flashinfer_backend.py:
```diff
@@ -278,18 +278,20 @@ def init_cuda_graph_state(
-            )
-        else:
-            cuda_graph_kv_indices = kv_indices_buf
+        if self.first_init == True:
```
(if we only init once, we can remove this if-first-init)

fzyzcjy commented on python/sglang/srt/layers/attention/flashinfer_backend.py:
```diff
@@ -408,6 +411,37 @@ def init_forward_metadata_capture_cuda_graph(
             )
             self.prefill_cuda_graph_metadata[bs] = prefill_wrappers
             self.forward_metadata = PrefillMetadata(prefill_wrappers, False, False)
+        elif forward_mode.is_extend():
```
question: it seems this branch is similar to the draft_extend branch. is it possible we somehow unify the two

(same for other `elif` branches in this file)

fzyzcjy commented on python/sglang/srt/model_executor/cuda_graph_runner.py:
```diff
@@ -162,6 +162,26 @@ def set_torch_compile_config():
 
+def get_prefill_seqlen_to_capture(model_runner: ModelRunner):
+    server_args = model_runner.server_args
+    if server_args.cuda_graph_prefill_max_seqlen is None:
+        capture_prefill_seqlen = list(range(32, 1025, 32))
+    else:
+        capture_prefill_seqlen = list(
+            range(32, server_args.cuda_graph_prefill_max_seqlen + 1, 32)
+        )
```
nit


```suggestion
    max_len = server_args.cuda_graph_prefill_max_seqlen or 1024
    capture_prefill_seqlen = list(range(32, max_len + 1, 32))
```

fzyzcjy commented on python/sglang/srt/model_executor/cuda_graph_runner.py:
```diff
@@ -162,6 +162,26 @@ def set_torch_compile_config():
+        )
+    compile_prefill_seqlen = (
+        [
+            # seqlen
+            # for seqlen in capture_prefill_seqlen
+            # if seqlen <= server_args.torch_compile_max_bs
```
hmm getting confused about this... do you mean we do not do any compile, then maybe just write

```
compile_prefill_seqlen = [] # unsupported currently
```

fzyzcjy commented on python/sglang/srt/model_executor/cuda_graph_runner.py:
```diff
@@ -263,11 +288,20 @@ def __init__(self, model_runner: ModelRunner):
         self.attn_tp_size = get_attention_tp_size()
         self.attn_tp_rank = get_attention_tp_rank()
 
+        self.max_num_token = 0
```
hmm it seems the field is not used globally, thus what about removing it and only have some local variables

fzyzcjy commented on python/sglang/srt/model_executor/cuda_graph_runner.py:
```diff
@@ -591,33 +895,82 @@ def capture_one_batch_size(self, bs: int, forward: Callable):
-            num_token_non_padded=self.num_token_non_padded,
+            num_token_non_padded=(
+                self.prefill_num_token_non_padded
+                if hasattr(self, "capture_forward_mode")
+                and self.capture_forward_mode == ForwardMode.EXTEND
```
nit: maybe make it unconditionally exist, then we can remove all "hasattr"

fzyzcjy commented on python/sglang/srt/model_executor/cuda_graph_runner.py:
```diff
@@ -591,33 +895,82 @@ def capture_one_batch_size(self, bs: int, forward: Callable):
+                else self.global_num_tokens_gpu
+            ),
+            global_num_tokens_for_logprob_gpu=(
+                self.prefill_global_num_tokens_for_logprob_gpu
+                if hasattr(self, "capture_forward_mode")
+                and self.capture_forward_mode == ForwardMode.EXTEND
+                else self.global_num_tokens_for_logprob_gpu
```
question: I see a pattern - we have some symmetric "blahblah" and "prefill_blahblah", then is it possible we extract it.

approach 1:

```
@dataclass
class Pack: # bad name, change it
  blahblah: Tensor
  another_thing: Tensor

pack = Pack(...)
prefill_pack = Pack(...)
```

then here we only need 

```
chosen_pack = prefill_pack if mode == EXTEND else pack

# several fields...
global_num_tokens_gpu=chosen_pack.global_num_tokens_gpu,
global_num_tokens_for_logprob_gpu=chosen_pack.global_num_tokens_for_logprob_gpu,
...
```

fzyzcjy replied:
another question: do you think it is better if we have one more GraphRunner tailored for prefill? Let's say, one way maybe BaseGraphRunner, DecodeGraphRunner, PrefillGraphRunner. Another way maybe, one class (GraphRunner) and two objects (prefill_runner = GraphRunner(); decode_runner = GraphRunner()).

also cc @merrymercy, since he (and another person iirc) is working on eagle cuda graph refactoring so I dnk whether will be related

fzyzcjy commented on python/sglang/srt/model_executor/cuda_graph_runner.py:
```diff
@@ -710,18 +1071,42 @@ def replay_prepare(
-        self.seq_lens[:raw_bs].copy_(forward_batch.seq_lens)
-        self.out_cache_loc[:raw_num_token].copy_(forward_batch.out_cache_loc)
-        self.positions[:raw_num_token].copy_(forward_batch.positions)
+        if forward_batch.forward_mode == ForwardMode.EXTEND:
+            self.prefill_input_ids[:raw_num_token].copy_(forward_batch.input_ids)
+            self.prefill_req_pool_indices[:raw_bs].copy_(forward_batch.req_pool_indices)
+            self.prefill_seq_lens[:raw_bs].copy_(forward_batch.seq_lens)
+            self.prefill_out_cache_loc[:raw_num_token].copy_(
+                forward_batch.out_cache_loc
+            )
+            self.prefill_positions[:raw_num_token].copy_(forward_batch.positions)
+        else:
+            self.input_ids[:raw_num_token].copy_(forward_batch.input_ids)
+            self.req_pool_indices[:raw_bs].copy_(forward_batch.req_pool_indices)
+            self.seq_lens[:raw_bs].copy_(forward_batch.seq_lens)
+            self.out_cache_loc[:raw_num_token].copy_(forward_batch.out_cache_loc)
+            self.positions[:raw_num_token].copy_(forward_batch.positions)
```
(this is another example of what I said above about the "Pack" approach or the "prefill_runner=GraphRunner();decode_runner=GraphRunner()" approach)

fzyzcjy reviewed (commented):
made a brief review

ZhouMengLei1999 commented:
> I will have a check tomorrow

I hope I can merge into main this time :)

ZhouMengLei1999 commented:
> > I will have a check tomorrow
> 
> I hope I can merge into main this time :)
After rebase main, something wrong, wait me to fix some bugs :( @fzyzcjy

Ximingwang-09 commented:
Great work! Is there any performance benchmark ?

ZhouMengLei1999 committed:
rebase main

ZhouMengLei1999 commented:
> > > > I will have a check tomorrow
> > > 
> > > 
> > > I hope I can merge into main this time :)
> > > After rebase main, something wrong, wait me to fix some bugs :( @fzyzcjy
> 
> I fixed this, but I need rebase again

I fixed this and rebase was done, please review again @fzyzcjy

fzyzcjy commented:
I will review now. But unfortunately usually a PR will not be merged very quickly, you can see even I myself have this many PRs not merged yet ---> https://github.com/sgl-project/sglang/issues?q=sort%3Aupdated-desc+is%3Aopen+is%3Apr+author%3A%40me :(

fzyzcjy commented:
ping me when the comments are solved and ready for review

ZhouMengLei1999 commented:
> > > you are welcome, feel free to ping me when the comments are solved and ready for review
> > 
> > 
> > @fzyzcjy I have resolved all the comments, please kindly review them again :)
> 
> Hello? @fzyzcjy Can you review this pr again? It's been four days since you last reviewed it :(

Can you reply to me? @fzyzcjy

ZhouMengLei1999 commented:
> > > > you are welcome, feel free to ping me when the comments are solved and ready for review
> > > 
> > > 
> > > @fzyzcjy I have resolved all the comments, please kindly review them again :)
> > 
> > 
> > Hello? @fzyzcjy Can you review this pr again? It's been four days since you last reviewed it :(
> 
> Can you reply to me? @fzyzcjy

@zhyncs @ispobock Can you find someone else to help review? I don't lknow what happened

ispobock commented:
Can you add benchmark results for the improvement and accuracy test results?
There are some related CI tests failed, pls have a check: 
https://github.com/sgl-project/sglang/actions/runs/17202869846/job/49229659880?pr=9186#step:4:1674

zhyncs commented:
@ZhouMengLei1999 please fix the conflicts. thanks
